app/uk/gov/hmrc/transitmovementsvalidator/services/JsonValidationService.scala

```diff
@@ -77,130 +70,40 @@ trait JsonValidationService extends ValidationService
-            case Failure(thr)                     => Future.successful(Left(Unexpected(Some(thr))))
-          }
+      val schemaValidator = schemaValidators(messageType.code)
+      validateJson(source, schemaValidator) match {
+        case Success(errors) if errors.isEmpty => Future.successful(Right(()))
+        case Success(errors) =>
+          val validationErrors = errors.map(
+            e => JsonSchemaValidationError(e.getSchemaPath, e.getMessage)
+          )
+
+          Future.successful(Left(JsonFailedValidation(NonEmptyList.fromListUnsafe(validationErrors.toList))))
+        case Failure(thr: JsonParseException) => Future.successful(Left(FailedToParse(stripSource(thr.getMessage))))
+        case Failure(thr)                     => Future.successful(Left(Unexpected(Some(thr))))
       }
     }
 
-  override def businessRuleValidation(messageType: String, source: Source[ByteString, _])(implicit
-    materializer: Materializer,
-    ec: ExecutionContext
-  ): EitherT[Future, ValidationError, Unit] =
-    EitherT {
-      MessageType.values.find(_.code == messageType) match {
-        case Some(_) =>
-          validateJson(source) match {
-            case Success(errors) if errors.isEmpty => Future.successful(Right(()))
-            case Success(errors)                   => Future.successful(Left(errors.head))
-            case Failure(thr)                      => Future.successful(Left(Unexpected(Some(thr))))
-          }
-        case None =>
-          Future.successful(Left(UnknownMessageType(messageType)))
-      }
-    }
-
-  //This method checks the type of message by inspecting the rootTag which appears to be a key in the JSON being validated
-  //It first splits the rootTag by ":" and uses the second part (at index 1) to find a matching MessageType from the set of all MessageType values.
-  //The equalsIgnoreCase method is used, meaning the match is case-insensitive.
-  //The function then checks what type of MessageType was found. If the message type is a subtype of DepartureMessageType
-  //and is one of the defined departure values, it returns the string "OfficeOfDeparture".
-  //If the message type is a subtype of ArrivalMessageType and is one of the defined arrival values, it returns the string "OfficeOfDestinationActual". For all other cases, it returns a None.
-  def checkMessageType(rootTag: String): Option[MessageType] = {
-    val messageType = MessageType.values.find(_.rootNode.equalsIgnoreCase(rootTag.split(":")(1)))
-    messageType
-  }
-
-  def validateJson(source: Source[ByteString, _], schemaValidator: JsonSchema)(implicit materializer: Materializer): Try[Set[ValidationMessage]] =
+  private def validateJson(source: Source[ByteString, _], schemaValidator: JsonSchema)(implicit materializer: Materializer): Try[Set[ValidationMessage]] =
     Using(source.runWith(StreamConverters.asInputStream(20.seconds))) {
```

Comment:
I know it's not part of this, but do you remember why we used 20 seconds?

Reply:
That's the standard timeout across HMRC services, including CTC, so if this takes longer than 20 seconds, the trader would have timed out anyway.